### PR TITLE
xAOD migration: Trigger

### PR DESCRIPTION
**Pull request by gerbaudo**
I believe @gerbaudo and @dantrim have made good progress on this. Are we all settled on this front such that w/ the new derivations w/ fixed xAOD information we'll be fine? Anything that needs to be looked at?

-s

### CODE REVIEW

**Comment by gerbaudo** [comment]
Attached `xaod_trigger` branch to this issue.
Note to self: used this command
```
curl --user "gerbaudo" --request POST --data '{"issue": "42", "head": "xaod_trigger", "base": "xaod"}' https://api.github.com/repos/gerbaudo/SusyCommon/pulls
```
from [http://stackoverflow.com/questions/4528869/how-do-you-attach-a-new-pull-request-to-an-existing-issue-on-github]

File: Root/SusyNtMaker.cxx [hunk 1]
```diff
@@ -584,8 +603,8 @@ void SusyNtMaker::storeJet(const xAOD::Jet &in)
     // jetOut->matchTruth    = m_isMC? matchTruthJet(jetIdx) : false;
 
     // B-tagging 
-    out.mv1           = (in.btagging())->MV1_discriminant();
-    out.sv1plusip3d   = (in.btagging())->SV1plusIP3D_discriminant();
+//    out.mv1           = (in.btagging())->MV1_discriminant();                   // dantrim - Feb 25 2015 - still causing seg-faults
```
Comment (gerbaudo) [inline Root/SusyNtMaker.cxx]: If this problem with `mv1` persists we should report it to the mailing list.

File: Root/XaodAnalysis.cxx [hunk 2]
```diff
@@ -869,8 +924,59 @@ bool XaodAnalysis::eleIsOfType(const xAOD::Electron &in, eleID id)
 void XaodAnalysis::fillEventTriggers()
 {
     if(m_dbg>=5) cout << "fillEventTriggers" << endl;
-
-    m_evtTrigFlags = 0;
+   
+    m_evtTrigFlags = 0; 
+
+    if(m_trigTool->isPassed("EF_e7_medium1"))                   m_evtTrigFlags |= triggerbits::TRIG_e7_medium1;
```
Comment (gerbaudo) [inline Root/XaodAnalysis.cxx]: Can we actually drop the `m_evtTrigFlags` completely, and only have the
implementation with `m_evtTriggerBits`? The `|=` is stil less clear than `SetBitNumber`
and having two "parallel" implementations (one with `long long` and one with `TBits`)
is confusing.

File: Root/SusyNtMaker.cxx [hunk 3]
```diff
@@ -1522,6 +1547,25 @@ struct FillCutFlow { ///< local function object to fill the cutflow histograms
     FillCutFlow& enableFilterNextCuts() { includeThisCut_ = true; return *this; }
 };
 //----------------------------------------------------------
+void SusyNtMaker::fillTriggerHisto() // dantrim trig
+{
+    for (int i = 0; i < triggerbits::N_TRIG; i++) {
+        if(m_trigTool->isPassed(triggerbits::trigger_names[i])) h_passTrigLevel->Fill(i+0.5);
```
Comment (gerbaudo) [inline Root/SusyNtMaker.cxx]: here can you use `m_evtTriggerBits.TestBitNumber()` instead?
(avoid calling `isPassed` multiple times)

File: SusyCommon/TriggerMap.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        BIT_xe80_tclcw_loose,
+        BIT_xe80T_tclcw_loose,
+        
+        
+        
+ 
+        N_TRIG
+    };
+    
+    // 2012 trigger bit masks
```
Comment (gerbaudo) [inline SusyCommon/TriggerMap.h]: these bit masks can be dropped with the `TBits` implementation

File: SusyCommon/TriggerMap.h [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+using namespace std;
+
+// --------------------------------------------------
+// TriggerMap
+//
+// A little module to hold anything related to the 
+// trigger bits and flags.
+// --------------------------------------------------
+/*
+struct triggerbits
```
Comment (gerbaudo) [inline SusyCommon/TriggerMap.h]: is this `triggerbits::bits` enum duplicated with the `triggerbits::enum` below?

**Comment by gerbaudo** [comment]
@dantrim thanks a lot for implementing this!! :-)
It looks good to me except for the few questions I posted.
When you think it's ready (even without object matching),
we should probably submit a few jobs and look again
at the data/mc plots. I can help with the job submission.

**Comment by dantrim** [comment]
Have updated to using TBits. In TriggerMap.h I have implemented the triggermap vector which removes redundancy and potential for error. Also is the version (commented out) of the "triggermap" using an map<string, int>. I think using the map<string, int> removes any possibility for calling/setting the incorrect bits (in future edits, etc..) as you need to call the bit from the map by name which will point to the associated bit whether or not the bit changes between productions. With the vector, the order still matters (as I have it now) since the bit is the index. Does this make sense?

File: SusyCommon/TriggerMap.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+#define SusyCommon_TriggerMap_h
+
+#include <iostream>
+#include <string>
+#include <map>
+
+
+#include "TBits.h"
+
+using namespace std;
```
Comment (gerbaudo) [inline SusyCommon/TriggerMap.h]: If possible, avoid using `using namespace` in header
http://stackoverflow.com/questions/5849457/using-namespace-in-c-headers

**Comment by gerbaudo** [comment]
Thank you! Looks good to me with the vector implementation.
I would just suggest to remove the chunks of code that are commented out,
and the `m_evtTrigFlags` datamember, which is now obsolete.
One more comment: it would be better to avoid defining too many
namespaces. For example, instead of defining
`namespace TriggerMap`, I would just use the existing `susy`:
```
namespace susy
{
const std::vector<std::string> triggermap = {
....
} // susy
```
(actually the triggermap is now a trigger list, you might want to rename it,
perhaps to smth like triggerNames)

**Comment by dantrim** [comment]
OK. I have made those changes. The TriggerMap namespace was remnant from when I was using the bit flags and did not want to get mixed up with those already defined in SusyNt.h. 

At some point, TriggerMap.h will have to go into SusyNtuple. Will also need to test/validate accessing the trigBits object stored in the susyNt.

The trigger tool is still unstable in the release I am using (2.0.21). When we move to more up-to-date releases (eg. 2.1.27) -- which also new SUSYTools-05-00-21 requires -- this is behavior is supposed to go away.

**Pull request merged**